src/snippet.rs

```diff
@@ -184,57 +263,77 @@ impl<'a> Annotation<'a> {
     }
 }
 
-/// Types of annotations.
-#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
-pub enum Level {
-    /// Error annotations are displayed using red color and "^" character.
-    Error,
-    /// Warning annotations are displayed using blue color and "-" character.
-    Warning,
-    Info,
-    Note,
-    Help,
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AnnotationKind {
+    Primary, // in rustc terms, "primary"; color to message's level
+    Context, // in rustc terms, "secondary"; fixed color
+             // Level(Level),  // from old API, dropped
 }
 
-impl Level {
-    pub fn title(self, title: &str) -> Message<'_> {
-        Message {
-            level: self,
-            id: None,
-            title,
-            snippets: vec![],
-            footer: vec![],
-        }
-    }
-
-    /// Create a [`Annotation`] with the given span for a [`Snippet`]
+impl AnnotationKind {
     pub fn span<'a>(self, span: Range<usize>) -> Annotation<'a> {
         Annotation {
             range: span,
             label: None,
-            level: self,
+            kind: self,
         }
     }
+
+    pub(crate) fn is_primary(&self) -> bool {
+        matches!(self, AnnotationKind::Primary)
+    }
 }
 
-impl Level {
-    pub(crate) fn as_str(&self) -> &'static str {
-        match self {
-            Level::Error => ERROR_TXT,
-            Level::Warning => WARNING_TXT,
-            Level::Info => INFO_TXT,
-            Level::Note => NOTE_TXT,
-            Level::Help => HELP_TXT,
-        }
+#[derive(Clone, Debug)]
+pub struct Origin<'a> {
+    pub(crate) origin: &'a str,
+    pub(crate) line: Option<usize>,
+    pub(crate) char_column: Option<usize>,
+    pub(crate) primary: bool,
+    pub(crate) standalone: bool,
+    pub(crate) label: Option<&'a str>,
+}
+
+impl<'a> From<Origin<'a>> for Element<'a> {
+    fn from(value: Origin<'a>) -> Self {
+        Element::Origin(value)
     }
+}
 
-    pub(crate) fn style(&self, stylesheet: &Stylesheet) -> Style {
-        match self {
-            Level::Error => stylesheet.error,
-            Level::Warning => stylesheet.warning,
-            Level::Info => stylesheet.info,
-            Level::Note => stylesheet.note,
-            Level::Help => stylesheet.help,
+impl<'a> Origin<'a> {
+    pub fn new(origin: &'a str) -> Self {
+        Self {
+            origin,
+            line: None,
+            char_column: None,
+            primary: false,
+            standalone: false,
+            label: None,
         }
     }
+
+    pub fn line(mut self, line: usize) -> Self {
+        self.line = Some(line);
+        self
+    }
+
+    pub fn char_column(mut self, char_column: usize) -> Self {
+        self.char_column = Some(char_column);
+        self
+    }
```

Comment:
These are new.  How do they relate to the rest of the API?

Should we make new parts of the API an extra commit on top of the existing API for easier comparison and  discussion?